queue/deadlineresolver.js

```diff
@@ -0,0 +1,201 @@
+var Promise       = require('promise');
+var debug         = require('debug')('queue:deadlineresolver');
+var slugid        = require('slugid');
+var assert        = require('assert');
+var _             = require('lodash');
+var base          = require('taskcluster-base');
+var data          = require('./data');
+var QueueService  = require('./queueservice');
+
+/** State that are considered resolved */
+const RESOLVED_STATES = [
+  'completed',
+  'failed',
+  'exception'
+];
+
+/**
+ * Facade that handles resolution tasks by deadline, using the advisory messages
+ * from the azure queue. The azure queue messages takes the form:
+ * `{taskId, deadline}`, and they become visible after `deadline` has been
+ * exceeded. The messages advice that if a task with the given `deadline` and
+ * `taskId` exists, then it be resolved by deadline, if not already resolved.
+ *
+ * Notice, that the task may not exists. Or the task may have different
+ * `deadline`, we shall only handle task if the `deadline` match.
+ */
+class DeadlineResolver {
+  /**
+   * Create DeadlineResolver instance.
+   *
+   * options:
+   * {
+   *   Task:           // instance of data.Task
+   *   queueService:   // instance of QueueService
+   *   publisher:      // publisher from base.Exchanges
+   *   pollingDelay:   // Number of ms to sleep between polling
+   *   parallelism:    // Number of polling loops to run in parallel
+   *                   // Each handles up to 32 messages in parallel
+   * }
+   */
+  constructor(options) {
+    assert(options, "options must be given");
+    assert(options.Task.prototype instanceof data.Task,
+           "Expected data.Task instance");
+    assert(options.queueService instanceof QueueService,
+           "Expected instance of QueueService");
+    assert(options.publisher, "Expected a publisher");
+    assert(typeof(options.pollingDelay) === 'number',
+           "Expected pollingDelay to be a number");
+    assert(typeof(options.parallelism) === 'number',
+           "Expected parallelism to be a number");
+    this.Task         = options.Task;
+    this.queueService = options.queueService;
+    this.publisher    = options.publisher;
+    this.pollingDelay = options.pollingDelay;
+    this.parallelism  = options.parallelism;
+
+    // Promise that polling is done
+    this.done         = null;
+    // Boolean that polling should stop
+    this.stopping     = false;
+  }
+
+  /** Start polling */
+  start() {
+    if (this.done) {
+      return;
+    }
+    this.stopping = false;
+
+    // Start a loop for the amount of parallelism desired
+    var loops = [];
+    for(var i = 0; i < this.parallelism; i++) {
+      loops.push(this.poll());
+    }
+    // Create promise that we're done looping
+    this.done = Promise.all(loops).catch((err) => {
```

Comment:
Like above, should we have some sort of fail safe that ensures that no individual loop gets jammed up?  It sounds like these messages are only available for a defined amount of time, so we should probably be careful not to let the queue spin for too long if one of these loops gets stuck